File: Project5/Solutions/cpp/dtree.cpp
#include "dtree.h"

#include <math.h>       /* log2 */
#include <queue>          // std::queue

/*
*   Dtree Constructor
*/
Dtree::Dtree() {
}



//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                              User Functions                              //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*   readInput:  Reads in inFilename and populates mdp data
*
*       Returns 0 on success, 1 on failure
*/
int Dtree::readInput() {
    std::ifstream inFile(inFilename);
    std::string line, val;

    // Read in the data
    if (inFile.is_open()) {
        // 1) Attributes
        getline(inFile, line);
        while (getline(inFile, line)) {
            // check if we finished the section
            if (line[0] == '%') {
                break;
            }

            // Find the Attribute Name
            std::size_t colon = line.find_first_of(':');
            string attribute = line.substr(0, colon);

            // Add the Attribute to our variables
            Attributes.push_back(attribute);
            attNameToInd[attribute] = numAttributes;
            attValues.push_back(vector<string>());

            // Add the Attribute Values to our variables
            line = line.substr(colon + 2);

            while (line.length() > 0) {
                std::size_t first_comma = line.find_first_of(',');
                if (first_comma == string::npos) {
                    attValues[numAttributes].push_back(line);
                    break;
                }
                attValues[numAttributes].push_back(line.substr(0, first_comma));
                line = line.substr(first_comma + 2);
            }

            // Increment # of Attributes
            numAttributes++;
        }

        // 2) Decision Values
        if (getline(inFile, line)) {
            while (line.length() > 0) {
                std::size_t first_comma = line.find_first_of(',');
                if (first_comma == string::npos) {
                    decisionVars.push_back(line);
                    decNameToInd[line] = numDecisionVars;
                    numDecisionVars++;
                    break;
                }
                decisionVars.push_back(line.substr(0, first_comma));
                decNameToInd[line.substr(0, first_comma)] = numDecisionVars;
                line = line.substr(first_comma + 2);
                numDecisionVars++;
            }
        } else { return 1; }

        // 3) Example Instances
        getline(inFile, line);
        getline(inFile, line);
        while (getline(inFile, line)) {
            // Create New Empty Example Vector
            Examples.push_back(vector<string>());

            // Populate the Example
            while (line.length() > 0) {
                std::size_t first_comma = line.find_first_of(',');
                if (first_comma == string::npos) {
                    Examples[numExamples].push_back(line);
                    break;
                }
                Examples[numExamples].push_back(line.substr(0, first_comma));
                line = line.substr(first_comma + 2);
            }

            // Increment the Number of Examples
            numExamples++;
        }
    } else {
        std::cout << "Error Opening Input File" << std::endl;
        return 1;
    }
    inFile.close();

    return 0;
}


/*
*   run
*/
void Dtree::run() {
    vector<int> attributes;
    for (int i=0; i < numAttributes; i++) {
        attributes.push_back(i);
    }

    decisionTree = decisionTreeLearning(Examples, attributes, Examples);
}


/*
*   writeOutput
*/
int Dtree::writeOutput() {
    std::ofstream outFile(outFilename);

    // BFS on tree
    std::queue<Node*> openList;
    openList.push(decisionTree);

    while (!openList.empty()) {
        // Grab Next Node off of the open list
        Node* currNode = openList.front(); openList.pop();

        // If its an Attribute Decision
        if (currNode->numChildren > 0) {
            // Print each edge to its children
            for (int i=0; i < currNode->numChildren; i++) {
                Node* child = currNode->children[i];

                // Print the line
                outFile << currNode->value << "? ";
                outFile << currNode->edges[i] << ", ";
                outFile << child->value;
                if (child->numChildren > 0) {
                    outFile << "?";
                }
                outFile << std::endl;

                // Add the child to the open list
                openList.push(child);
            }
        }
    }

    outFile.close();
    return 0;
}

//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                    Decision Tree Learning Functions                      //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*
*/
Node* Dtree::decisionTreeLearning(vector<vector<string>> examples,
                                 vector<int> attributes,
                                 vector<vector<string>> parent_examples) {
    Node* tree;

    if (examples.empty()) {
        return pluralityValue(parent_examples);
    } else if (sameClassification(examples)) {
        return pluralityValue(examples);            // the classification
    } else if (attributes.empty()) {
        return pluralityValue(examples);
    } else {
        // vector<string> A = maxImportance(attributes, examples);
        int A = maxImportance(attributes, examples);

        // A new decision tree with root test A
        tree = new Node(Attributes[A]);

        // Next Attributes (erase curr maxImportance)
        vector<int> nextAttr = attributes;
        for (int i=0; i < nextAttr.size(); i++) {
            if (nextAttr[i] == A) {
                nextAttr.erase(nextAttr.begin() + i);
            }
        }

        for (int i=0; i < attValues[A].size(); i++) {
            // Compute exs
            vector<vector<string>> exs;
            for (int j=0; j < examples.size(); j++) {
                if (examples[j][A] == attValues[A][i]) {
                    exs.push_back(examples[j]);
                }
            }

            // Recursively Call dTL to populate subtrees
            Node* subtree = decisionTreeLearning(exs, nextAttr, examples);

            // Add a branch to the tree
            tree->addChild(subtree, attValues[A][i]);
        }
    }

    // Return the decision tree
    return tree;
}

//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                    Decision Tree Helper Functions                        //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*
*/
Node* Dtree::pluralityValue(vector<vector<string>> examples) {
    vector<int> values(numDecisionVars, 0);

    // Tally up # of values
    for (int i=0; i < examples.size(); i++) {
        string value = examples[i][numAttributes];
        values[decNameToInd[value]]++;
    }

    // Find the plurality value (max # of tallies)
    int maxNum = 0;
    int maxVal = 0;
    for (int i=0; i < values.size(); i++) {
        if (values[i] > maxNum) {
            maxNum = values[i];
            maxVal = i;
        }
    }

    // Construct a subtree and return it
    Node* tree = new Node(decisionVars[maxVal]);
    return tree;
}


/*
*   sameClassification
*
*       Returns 1 if all examples have the same classification
*               0 otherwise
*/
bool Dtree::sameClassification(vector<vector<string>> examples) {
    string value = examples[0][numAttributes];

    for (int i=1; i < examples.size(); i++) {
        if (value != examples[i][numAttributes]) {
            return 0;
        }
    }

    return 1;
}


/*
*   maxImportance
*
*       Returns the attribute of max importance
*/
int Dtree::maxImportance(vector<int> attributes,
                         vector<vector<string>> examples) {
    // Compute first attribute's importance for comparison
    double maxValue = importance_Gen(attributes[0], examples);
    int maxAttribute = attributes[0];

    // Loop over all others, keeping track of the most important attribute
    for (int i=1; i < attributes.size(); i++) {
        double nextValue = importance_Gen(attributes[i], examples);
        if (nextValue > maxValue) {
            maxValue = nextValue;
            maxAttribute = attributes[i];
        }
    }

    // return the most important attribute
    return maxAttribute;
}

//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                    maxImportance Helper Functions                        //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*   importance
*       We assume that there are only 2 decision variables
*
*       Returns the importance of the given attribute on the examples 
*/
double Dtree::importance(int attribute, vector<vector<string>> examples) {
    // Calculate p & n
    string yes = decisionVars[0];
    int p = 0;
    for (int i=0; i < examples.size(); i++) {
        if (examples[i][numAttributes] == yes) {
            p++;
        }
    }
    int n = examples.size() - p;

    // Calculate the importance
    // double importanceValue = B(p,n) - Remainder(attribute, examples, p, n);

    // Calculate the importance
    double B_Val = B(p, n);
    double R_Val = Remainder(attribute, examples, p, n);
    double importanceValue = B_Val - R_Val;

    /*
    std::cout << "Regular Function" << std::endl;
    std::cout << "attribute = " << attribute << std::endl;
    std::cout << "B_Val = " << B_Val << std::endl;
    std::cout << "R_Val = " << R_Val << std::endl;
    std::cout << "importanceValue = " << importanceValue;
    std::cout << std::endl;
    */

    return importanceValue;
}


/*
*   B
*       Returns the entropy of a boolean R.V. with prob q
*/
double Dtree::B(int p, int n) {
    // Catch 0-entropy case to avoid NANs
    if ( (p == 0) || (n == 0) ) {
        return 0;
    }

    double q = static_cast<double>(p / (p+n));
    double entropy = -1 * (q*log2(q) + (1-q)*log2(1-q));

    return entropy;
}


/*
*   Remainder
*/
double Dtree::Remainder(int attr,
                        vector<vector<string>> examples,
                        int p,
                        int n) {
    double value = 0;

    for (int k=0; k < attValues[attr].size(); k++) {
        // Calculate pk and nk
        int pk = 0;
        int nk = 0;

        // Loop over all of the examples
        for (int j=0; j < examples.size(); j++) {
            // for every example with the attr value, calculate pk/nk
            if (examples[j][attr] == attValues[attr][k]) {
                if (examples[j][numAttributes] == decisionVars[0]) {
                    pk++;
                } else {
                    nk++;
                }
            }
        }

        if ((pk + nk) == 0) {
            continue;
        } else {
            // Calculate term for this k
            double firstPart = static_cast<double> ((pk+nk)/(p+n));
            value += firstPart * B(pk, nk);
        }
    }

    // return the value
    return value;
}
//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//            maxImportance Helper Functions   (General Case)               //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*   importance
*       We assume that there are only 2 decision variables
*
*       Returns the importance of the given attribute on the examples 
*/
double Dtree::importance_Gen(int attribute, vector<vector<string>> examples) {
    // 1) Calculate occurances of each decision variables
    vector<double> varCount(numDecisionVars, 0);
    double totalCount = 0;
    for (int i=0; i < numDecisionVars; i++) {
        for (int j=0; j < examples.size(); j++) {
            if (examples[j][numAttributes] == decisionVars[i]) {
                varCount[i]++;
                totalCount++;
            }
        }
    }

    // 1.5) If totalCount == 0, quit
    if (totalCount == 0) {
        return 0;
    }

    // 2) Normalize to obtain probabilities
    vector<double> varProb(numDecisionVars, 0);
    for (int i=0; i < numDecisionVars; i++) {
        varProb[i] = varCount[i] / totalCount;
    }

    // 3) Calculate the importance
    double B_Val = B_Gen(varProb);
    double R_Val = Remainder_Gen(attribute, examples, varProb, totalCount);
    double importanceValue = B_Val - R_Val;

    /*
    std::cout << "  General Function" << std::endl;
    std::cout << "attribute = " << attribute << std::endl;
    std::cout << "B_Val = " << B_Val << std::endl;
    std::cout << "R_Val = " << R_Val << std::endl;
    std::cout << "importanceValue = " << importanceValue << std::endl;
    std::cout << std::endl;
    */

    return importanceValue;
}


/*
*   B_Gen
*       Returns the entropy of any R.V. with probs in vector
*/
double Dtree::B_Gen(vector<double> vals) {
    double entropy = 0;
    for (int i=0; i < vals.size(); i++) {
        if (vals[i] == 0) {
            continue;
        }
        entropy += -1 * vals[i] * log2(vals[i]);
    }

    return entropy;
}

/*
*   Remainder_Gen
*/
double Dtree::Remainder_Gen(int attr,
                            vector<vector<string>> examples,
                            vector<double> varProb,
                            double totalCount) {
    double value = 0;

    for (int k=0; k < attValues[attr].size(); k++) {
        vector<double> vals(numDecisionVars, 0);
        double total = 0;

        // 1) Count Occurences of specified Attribute for each output
        for (int j=0; j < examples.size(); j++) {
            if (examples[j][attr] == attValues[attr][k]) {
                for (int i=0; i < numDecisionVars; i++) {
                    if (examples[j][numAttributes] == decisionVars[i]) {
                        vals[i]++;
                        total++;
                    }
                }
            }
        }

        // 1.5) Check for dividing by zero
        if (total == 0) {
            continue;
        }

        // 2) Normalize
        vector<double> probs(numDecisionVars, 0);
        for (int i=0; i < numDecisionVars; i++) {
            probs[i] = vals[i] / total;
        }

        // 3) Calculate term for this k
        value += total / totalCount * B_Gen(probs);
    }

    // return the value
    return value;
}

File: Project #5/Solutions/python/dtree.py
""" Problem 4 Decision Tree Generation """

import csv
import copy
import math
from collections import deque


class Queue:
    """ FIFO Queue class """
    def __init__(self):
        self.queue = deque()

    def push(self, item):
        """ Push items onto queue """
        self.queue.append(item)

    def pop(self):
        """ Pop items off of queue """
        return self.queue.popleft()

    def is_empty(self):
        """ empty check """
        return self.size() == 0

    def size(self):
        """ returns size of queue """
        return len(self.queue)

class Node:
    """ Node of a tree """
    def __init__(self, value):
        self.value = value
        self.children = []
        self.edges = []

    def add_child(self, child_node, edge):
        """ Add a child node """
        self.children.append(child_node)
        self.edges.append(edge)

class DecisionTree:
    """ Decision Tree Generation Algorithm Class """
    def __init__(self):
        self.attributes = []
        self.att_idxs = {}
        self.att_values = {}
        self.dec_values = []
        self.examples = []
        self.dtree = None

    def read_dtree_file(self, fname="examples.txt"):
        """ Reads in examples from file """
        with open(fname, 'r') as infile:
            csvreader = csv.reader(infile)
            lines = []
            for line in csvreader:
                lines.append(line)

            # 1) Grab the attributes and their possible values (index 1)
            curr_line_idx = 1
            for line_idx, next_line in enumerate(lines[1:]):
                # Check if we're done
                if next_line[0][0] == "%":
                    curr_line_idx += line_idx
                    break

                # Grab the attribute name and first option
                colon_ind = next_line[0].rfind(":")
                att_name = next_line[0][:colon_ind]
                self.attributes.append(att_name)
                self.att_idxs[att_name] = len(self.attributes) - 1

                self.att_values[att_name] = []
                self.att_values[att_name].append(next_line[0][colon_ind+1:].strip())

                # Grab the remaining options
                for att_val in next_line[1:]:
                    self.att_values[att_name].append(att_val.strip())

            # 2) Grab the decision values
            curr_line_idx += 1
            for dec_value in lines[curr_line_idx]:
                self.dec_values.append(dec_value.strip())

            # 3) Grab the example instances
            curr_line_idx += 3
            for ex_line in lines[curr_line_idx:]:
                # Check if we're done or there was a mismatch
                if len(ex_line) != (len(self.attributes) + 1):
                    break

                example = []
                for value in ex_line:
                    example.append(value.strip())
                self.examples.append(example)

    def run(self):
        """ Run the decision tree learning algorithm on the example data """
        self.dtree = self.decision_tree_learning(self.examples, self.attributes, self.examples)

    def decision_tree_learning(self, examples, attributes, parent_examples):
        """ Recursive decision tree learning algorithm """

        if len(examples) == 0:
            return self.plurality_value(parent_examples)
        elif self.same_classification(examples):
            return self.plurality_value(examples)
        elif len(attributes) == 0:
            return self.plurality_value(examples)
        else:
            next_attr = self.max_importance(attributes, examples)
            tree = Node(next_attr)

            for attr_value in self.att_values[next_attr]:
                # Collect examples in which next_attr has value attr_value
                exs = []
                for example in examples:
                    if example[self.att_idxs[next_attr]] == attr_value:
                        example_copy = copy.deepcopy(example)
                        exs.append(example_copy)

                # Remove next_attr from attributes
                next_attributes = copy.deepcopy(self.attributes)
                next_attributes.remove(next_attr)

                # Recursively call decision_tree_learning to populate subtrees
                sub_tree = self.decision_tree_learning(exs, next_attributes, examples)

                # Add a branch to the tree
                tree.add_child(sub_tree, attr_value)

            return tree

    def plurality_value(self, examples):
        """ Returns the plurality value for the given examples """
        # 1) Tally up # of values
        dec_value_counts = [0] * len(self.dec_values)
        for idx, dec_value in enumerate(self.dec_values):
            for example in examples:
                if example[-1] == dec_value:
                    dec_value_counts[idx] += 1

        # 2) Find the plurality value
        max_count = dec_value_counts[0]
        max_value = self.dec_values[0]
        for idx, dec_value in enumerate(self.dec_values):
            if dec_value_counts[idx] > max_count:
                max_count = dec_value_counts[idx]
                max_value = dec_value

        # 3) Construct a subtree and return it
        tree = Node(max_value)
        return tree


    def same_classification(self, examples):
        """ Returns true if all examples have the same classification, false otherwise """
        value = examples[0][-1]
        for example in examples[1:]:
            if value != example[-1]:
                return False
        return True

    def max_importance(self, attributes, examples):
        """ Returns the max importance attribute for given examples """
        # 1) Compute first elements importance for comparison
        max_value = self.importance(attributes[0], examples)
        max_attr = attributes[0]

        # 2) Loop over all others, keeping track of the most important attribute
        for attribute in attributes:
            next_value = self.importance(attribute, examples)
            if next_value > max_value:
                max_value = next_value
                max_attr = attribute

        return max_attr

    def importance(self, attribute, examples):
        """ Returns the importance of an attribute given examples """
        # 1) Calculate the occurences of each decision value
        dec_value_counts = [0] * len(self.dec_values)
        for idx, dec_value in enumerate(self.dec_values):
            for example in examples:
                if example[-1] == dec_value:
                    dec_value_counts[idx] += 1

        # 1.5) If no matches, quit
        total_count = sum(dec_value_counts)
        if total_count == 0:
            return 0

        # 2) Normalize to obtain probabilities
        dec_value_probs = [0.0] * len(self.dec_values)
        for i in range(0, len(self.dec_values)):
            dec_value_probs[i] = dec_value_counts[i] / total_count

        # 3) Calculate the importance
        b_val = self.b_calc(dec_value_probs)
        r_val = self.remainder_calc(attribute, examples, dec_value_probs, total_count)
        importance_value = b_val - r_val

        return importance_value

    def b_calc(self, dec_value_probs):
        """ Returns the entropy of and R.V. with probs in list """
        entropy = 0
        for dec_value_prob in dec_value_probs:
            if dec_value_prob != 0:
                entropy += -1 * dec_value_prob * math.log2(dec_value_prob)
        return entropy


    def remainder_calc(self, attribute, examples, dec_value_probs, total_count):
        """ Calculates remainder """
        value = 0

        # Loop over the attribute values
        for att_value in self.att_values[attribute]:
            vals = [0] * len(self.dec_values)
            total = 0

            # 1) Count occurences of specified attribute for each output
            for example in examples:
                if example[self.att_idxs[attribute]] == att_value:
                    for dec_idx, dec_value in enumerate(self.dec_values):
                        if example[-1] == dec_value:
                            vals[dec_idx] += 1
                            total += 1

            # 1.5) Check for dividing by zero
            if total == 0:
                continue

            # 2) Normalize
            probs = [0.0] * len(self.dec_values)
            for i in range(0, len(self.dec_values)):
                probs[i] = vals[i] / total

            # 3) Calculate term for this k
            value += total / total_count * self.b_calc(probs)

        return value

    def write_output(self, fname="dtree.txt"):
        """ Write the output to a file """

        # 1) Open the file for writing
        f = open(fname, "w")

        # 2) Construct the output string (BFS on tree)
        out_str = ""
        open_list = Queue()
        open_list.push(self.dtree)

        while not open_list.is_empty():
            # Grab the next node off of the open list
            curr_node = open_list.pop()

            # If it's an attribute decision, process it
            if len(curr_node.children) > 0:
                # And loop over each of its children
                for child_idx, child in enumerate(curr_node.children):
                    # Printing a line for each child
                    out_str += curr_node.value + "? "
                    out_str += curr_node.edges[child_idx] + ", "
                    out_str += child.value
                    if len(child.children) > 0:
                        out_str += "?"
                    out_str += "\n"

                    # And adding the children that are parents themselves to the open list
                    open_list.push(child)

        # Remove last new line
        # out_str = out_str[:-1]

        # 3) Write to the output file and then close it
        f.write(out_str)
        f.close()




def main():
    """ main function for calling dtree generation algorithm """
    my_dtree = DecisionTree()

    # 1) Read in the input file
    my_dtree.read_dtree_file()

    # 2) Run decision tree learning
    my_dtree.run()

    # 3) Write output file
    my_dtree.write_output()

if __name__ == "__main__":
    main()
